### PR TITLE
Refactoring of the sct_resample to separate API and command-line tool.

### PR DESCRIPTION
**Pull request by perone**
This is a refactoring of the `sct_resample` command-line tool, according to the new API/command-line tool design.

Changes in this PR:
- Moved the `sct_resample` logic code to a new module called `resample` inside the `spinalcordtoolbox` module;
- The `sct_resample` from the scripts folder is now a wrapper around the API;
- In the API we have now two functions called `resample_image` and `resample_file`, the first one will resample a nifti image object and the later will resample from an input file and will output a new file;
- Didn't change the original resampling code;
- `sct_testing` tested it without any issues and I also tested on other images;

image

### CODE REVIEW

File: scripts/sct_resample.py [hunk 1]
```diff
@@ -1,31 +1,24 @@
-#!/usr/bin/env python
 #########################################################################################
 #
 # Resample data.
 #
 # ---------------------------------------------------------------------------------------
 # Copyright (c) 2014 Polytechnique Montreal <www.neuro.polymtl.ca>
 # Authors: Julien Cohen-Adad, Sara Dupont
-# Modified: 2015-09-08
+# Modified: 2017-11-22 (moved the code to spinalcordtoolbox module)
```
Comment (zougloub) [inline scripts/sct_resample.py]: The change log is contained inside git history, does it make sense to keep these lines?

File: scripts/sct_resample.py [hunk 2]
```diff
@@ -1,31 +1,24 @@
...
 import commands
 import sct_utils as sct
-# import time
-# from sct_convert import convert
-from msct_image import Image
+
 from msct_parser import Parser
-# import dipy.align.reslice as dp_iso
 
+from spinalcordtoolbox.resample import nipy_resample
```
Comment (zougloub) [inline scripts/sct_resample.py]: Unless the function is used dozens of times, it's recommended to keep its fully qualified path.

File: scripts/sct_resample.py [hunk 3]
```diff
@@ -1,31 +1,24 @@
...
 
-# TODO: adjust qform of output data to account for pixel size
-# TODO: test if crashes with 2d or 4d data
-# TODO: raise exception if input size is not numerical
-
 import sys
-# import os
-# import getopt
+
 import commands
```
Comment (zougloub) [inline scripts/sct_resample.py]: Unused

File: spinalcordtoolbox/resample/nipy_resample.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    elif interpolation == 'linear':
+        interp_order = 1
+    elif interpolation == 'spline':
+        interp_order = 2
+
+    # create 3d coordmap because resample only accepts 3d data (jcohenadad 2016-07-26)
+    if len(n) == 4:
+        coordmap3d = deepcopy(input_image.coordmap)
+        from nipy.core.reference.coordinate_system import CoordinateSystem
+        coordmap3d.__setattr__('function_domain', CoordinateSystem('xyz'))
```
Comment (zougloub) [inline spinalcordtoolbox/resample/nipy_resample.py]: `coordmap3d.function_domain = Coord...` ?

File: spinalcordtoolbox/resample/nipy_resample.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+import numpy as np
+from nipy.algorithms.registration.resample import resample as n_resample
+
+def resample_image(input_image, new_size, new_size_type,
+                   interpolation='linear', verbose=1):
+    """This function will resample the specified input
+    image to the target size.
+
+    :param input_image: The input image.
+    :param new_size: The target size, i.e. 0.25x0.25
```
Comment (zougloub) [inline spinalcordtoolbox/resample/nipy_resample.py]: If `0.25x0.25` is a string, put in quotes to be clearer wrt/ param type

File: spinalcordtoolbox/resample/nipy_resample.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
+#########################################################################################
+#
+# Resample data.
```
Comment (zougloub) [inline spinalcordtoolbox/resample/nipy_resample.py]: Could indicate that this is using nipy here.

File: spinalcordtoolbox/resample/nipy_resample.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+# ---------------------------------------------------------------------------------------
+# Copyright (c) 2014 Polytechnique Montreal <www.neuro.polymtl.ca>
+# Authors: Julien Cohen-Adad, Sara Dupont
+# Modified: 2017-11-22 (moved the code to spinalcordtoolbox module)
+#
+#
+# About the license: see the file LICENSE.TXT
+#########################################################################################
+import sys
+import commands
```
Comment (zougloub) [inline spinalcordtoolbox/resample/nipy_resample.py]: Unused?

File: spinalcordtoolbox/resample/nipy_resample.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+from copy import deepcopy
+
+import sct_utils as sct
```
Comment (zougloub) [inline spinalcordtoolbox/resample/nipy_resample.py]: You could rework the order of imports to [batteries included, 3rd party, internal] while you're at it.

File: spinalcordtoolbox/resample/nipy_resample.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+# Copyright (c) 2014 Polytechnique Montreal <www.neuro.polymtl.ca>
+# Authors: Julien Cohen-Adad, Sara Dupont
+# Modified: 2017-11-22 (moved the code to spinalcordtoolbox module)
+#
+#
+# About the license: see the file LICENSE.TXT
+#########################################################################################
+import sys
+import commands
+from copy import deepcopy
```
Comment (zougloub) [inline spinalcordtoolbox/resample/nipy_resample.py]: Unless used 10 times... same thing.

File: scripts/sct_resample.py (hunk 1 above)
Comment (jcohenadad) [reply scripts/sct_resample.py]: no we can remove

**Comment by perone** [comment]
I tried to avoid touching the code (I basically moved it to the new design), but I can fix the issues and then update the PR.

File: scripts/sct_resample.py (hunk 1 above)
Comment (perone) [reply scripts/sct_resample.py]: Removed.

File: scripts/sct_resample.py (hunk 3 above)
Comment (perone) [reply scripts/sct_resample.py]: which one ? the `sys` and `commands` are being used

File: scripts/sct_resample.py (hunk 2 above)
Comment (perone) [reply scripts/sct_resample.py]: You mean using: "spinalcordtoolbox.resample.nipy_resample(...)" in the call ?

File: spinalcordtoolbox/resample/nipy_resample.py (hunk 6 above)
Comment (perone) [reply spinalcordtoolbox/resample/nipy_resample.py]: Added a comment.

File: spinalcordtoolbox/resample/nipy_resample.py (hunk 7 above)
Comment (perone) [reply spinalcordtoolbox/resample/nipy_resample.py]: Removed.

File: spinalcordtoolbox/resample/nipy_resample.py (hunk 8 above)
Comment (perone) [reply spinalcordtoolbox/resample/nipy_resample.py]: Done

File: spinalcordtoolbox/resample/nipy_resample.py (hunk 9 above)
Comment (perone) [reply spinalcordtoolbox/resample/nipy_resample.py]: Done

File: spinalcordtoolbox/resample/nipy_resample.py (hunk 5 above)
Comment (perone) [reply spinalcordtoolbox/resample/nipy_resample.py]: Added quotes.

File: spinalcordtoolbox/resample/nipy_resample.py (hunk 4 above)
Comment (perone) [reply spinalcordtoolbox/resample/nipy_resample.py]: Done

File: spinalcordtoolbox/resample/nipy_resample.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+# About the license: see the file LICENSE.TXT
+#########################################################################################
+import copy
+
+import nipy
+import numpy as np
+from nipy.algorithms.registration.resample import resample as n_resample
+
+import sct_utils as sct
+from msct_image import Image
```
Comment (peristeri) [inline spinalcordtoolbox/resample/nipy_resample.py]: I don't think you are using `Image` or `Parser` anywhere in the code.

Comment (perone) [reply spinalcordtoolbox/resample/nipy_resample.py]: Fixed.

**Comment by zougloub** [comment]
Now that the branch applies the right changes, do we want to have it cleaned-up (interactive rebase) before it's merged, or taken it as is?

**Comment by peristeri** [comment]
@zougloub you can squash and merge.

**Comment by zougloub** [comment]
@perone sorry to bother but you have indented with tabs, not spaces, in hunk ```
@@ -127,11 +125,11 @@ def run_main():
```
I'd merge after you fix it (if you can do a `git rebase -i` and rework the branch, it will be even better ;).

**Comment by zougloub** [comment]
Turns out that since spinalcordtoolbox is in `$SCT_DIR` which is never added to `$PYTHONPATH`, the `import spinalcordtoolbox.[...]` may fail. Can we add it to the launcher @perone ?

**Pull request merged**